### DISCUSSION TITLE
Ultra-lightweight firmware with help of external server

### DISCUSSION THREAD

**Discussion by korniux** [xblax/flashforge_ad5m_klipper_mod#260] [Ideas]
Got 'mcu' memory failure error too often (after >200g of printing), so I connected to printer and looked like most of RAM consumed ин moonraker. Is it possible:
- forward socket with socat or SSH of klipper
- deploy moonraker on external server (ie NUC, pi or zima) and connect socket to it

**Comment by KaruroChori** [comment]
Technically yes, but this will likely result in even more timeout issues.
If I remember correctly this idea was tested very briefly before the public release of the mod but the outcome was not usable.

Flashforge does not use moonraker in their semi-custom implementation, so they never designed this hardware to fit it in.

We might get around that (as long as there is not an actual leak) by tweaking the parameters for the garbage collection in python for moonraker.

**Comment by xblax** [comment]
I tested very-early on to connect Moonraker to the built-in Klipper from Flagshforge. On reason why I abandoned that Idea, is that Moonraker expects to run on the same machine as Klipper with shared file system access, for example the uploaded G-Code is saved into a folder by Moonraker and klippy accesses it directly. I gives much neater integration that way.

It might work to have both Klipper and Moonraker in a spearate PI and only tunnel the MCU serial connections via SSH, but didn't test it. Latency over SSH could add new problems. It makes other things more complicated like using the display, buzzer etc.

I myself won't do further investigation in this direction, because for me it mostly works fine (and also didn't have much time at all in the last months). 

I use the mod myself mostly with Klipper-Screen, direct upload and print start from Prusa Slicer and don't really need Fluidd/Mainsail while printing - didn't have the issue in a long time, but the parts I print are also mostly smaller (<100-150g Filament).
